concepts/http-server/about.md

```diff
@@ -0,0 +1,108 @@
+"Serve accepts incoming HTTP connections on the listener l, creating a new service goroutine for each. The service goroutines read requests and then call handler to reply to them.". Therefore each connection has its own goroutine which enables go to handle client requests concurrently. Albeit, this leads to better performance especially for handling a large set of clients, you must be aware of shared access to a resource and use mutexes or channels to handle it.
+~~~~
+
+### Handlers
+
+```go
+type Handler interface {
+ ServeHTTP(ResponseWriter, *Request)
+}
+```
+
+Handlers are responsible for satisfying the client's requests by performing various tasks. Each object that implements the `http.Handler` interface, is capable of responding to client requests. Handler functions receive an `http.Request` which contains information regarding the client’s request, and an `http.ResponseWriter` that allows the handler to write the appropriate response in accordance with the request. `http.ResponseWriter` interface contains a `Header` field (!!! I need to complete this !!!), a `Write([]byte) (int, error)` method that writes the response to the client, and a `WriteHeader(statusCode int)`  that writes the appropriate status code to the client.
+
+~~~~exercism/caution
+If you wish to send and status code besides `http.StatusOk` you need to use `WriteHeader` before `Write`, since `Write` assumes `http.StatusOk` if no status is set before invoking it.  
+Example:  
+```go
+func(w http.ResponseWriter, r *http.Request){
+ w.Write([]byte("Not Found!"))
+ w.WriteHeader(http.StatusNotFound)
+}
+// The result of running this function is a response with status code 200 Ok!
+----------
+func(w http.ResponseWriter, r *http.Request){
+ w.WriteHeader(http.StatusNotFound)
+ w.Write([]byte("Not Found!"))
+}
+// The result of running this function is a response with status code 404 NotFound!
+```
+~~~~
+
+### Multiplexer
+
+`http.ServeMux` is responsible for routing each request to the right handler by matching the pattern in the request with the **longest** pattern specified for it. `http.NewServeMux() *ServeMux` is used to create a new multiplexer. It uses handlers to route each request to the proper handler. As an example:
+```go
+func multiplexer() *ServeMux{
+ serveMux := http.NewServeMux()
+ serveMux := http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request){
+ w.Write([]byte("Hello World!"))
+ })
+ return serveMux
+}
+```
+
+#### Absolute paths vs subtrees
+
+Paths that end without a `/` are called absolute paths (e.g. `"root/image"`). If an absolute path is used in the request, the multiplexer will try to find the “exact” match for that path. On the contrary, paths that end with `/` are called subtrees (e.g. `“root/image/“`) and the multiplexer will try to find the best match for these paths. Beware that the multiplexer will convert an absolute path to a subtree if it does not find a corresponding handler.  
```

Comment:
I am not familiar with this. What does "best match" mean here? And isn't the solution to this to add middleware that always removes the trailing slash so this never happens?

Reply:
For example if you have root/image/a/ and a/ doesn't exist, it will send it to the handler with root/image/.
Yes that is the solution to remove the trailing slash, but I thought it is important to mention this behavior.